Add module callbacks called for reacting to deactivation and profile update (#12062)
matrix-org/synapse: docs/modules/third_party_rules_callbacks.md

If multiple modules implement this callback, Synapse runs them all in order.

> ### `on_profile_update`

Reviewer: Is it possible to distinguish in this method whether a user was deactivated versus simply setting their profile information to a display name of `""` and an avatar URL of `None`? I could see implementations of this method commonly having a conditional of:

```python
if new_profile.display_name == "" and new_profile.avatar_url is None:
    return
```

in which case users would be able to trigger a profile update without this code catching it.

Can we track whether a profile update is the result of a deactivation and refrain from running this method in that instance?

Author: > Can we track whether a profile update is the result of a deactivation and refrain from running this method in that instance?

I don't think we should skip this method in this case. It sounds to me like a way to get modules out of sync with a user's profile (e.g. if they maintain their own db table for profiles, and the user gets reactivated and changes just their display name, the module will still think they've got their old avatar).
We could pass a flag onto the profile handler then to the callback. The initial intention was to have modules react to deactivation separately in `on_deactivation`, but I guess it doesn't hurt to have this information available there as well.

Author: (done in https://github.com/matrix-org/synapse/pull/12062/commits/698afcf29a50f644423d587bbb04923249ecd9db)

_First introduced in Synapse v1.5X.0_

```python
async def on_profile_update(
    user_id: str,
    new_profile: "synapse.module_api.ProfileInfo",
    by_admin: bool,
) -> None:
```

> Called after updating a local user's profile. The update can be triggered either by the

Reviewer: Could you also note here that per-room profile updates do not trigger this callback?

user themselves or a server admin. The update can also be triggered by a user being
deactivated (in which case their display name is set to an empty string (`""`) and the

> avatar URL is set to `None`). The module is passed the Matrix ID of the user whose profile

Reviewer: Could you note that this method will not be called when a user sets their global profile information for the first time upon registration?

has been updated, their new profile, as well as a boolean that is `True` if the update
was triggered by a server admin (and `False` otherwise). Note that this boolean is also

> `True` if the profile change happens as a result of the user logging through Single
> Sing-On, and if a server admin updates their own profile.

Reviewer: ```suggestion
`True` if the profile change happens as a result of the user logging in through Single
Sign-On, or if a server admin updates their own profile.
```

If multiple modules implement this callback, Synapse runs them all in order.

### `on_deactivation`

_First introduced in Synapse v1.5X.0_

```python

> async def on_deactivation(user_id: str, by_admin: bool) -> None:

Reviewer: Note that we officially support re-activating users, which currently modules would not be told about: https://github.com/matrix-org/synapse/blob/18862f20b5495bdc556c54e92fd4b1efdc718ba7/synapse/rest/admin/users.py#L333-L335

It might be worthwhile to modify this callback to add support for information modules about re-activated users?

Author: I wonder if it's not better to add a `on_reactivation` callback in the future (when we end up needing one)? It feels more like the right way to convey this information to modules.

Reviewer: Compiling both into one method would have the advantage of reducing the number of module callbacks we have, while potentially having the downside of a slightly more confusing function signature (`by_admin` would always be `True` on reactivation).

So I suppose it depends on how much value we put on keeping the callback surface small. If not much, then having both `on_deactivation` and `on_reactivation` methods are fine.

Small nit: `on_deactivation` -> `on_user_deactivation`

the update is a result of the user being deactivated. Note that the `by_admin` boolean is
also `True` if the profile change happens as a result of the user logging in through
Single Sign-On, or if a server admin updates their own profile.

Per-room profile changes do not trigger this callback to be called. Synapse administrators
wishing this callback to be called on every profile change are encouraged to disable
per-room profile globally using the `allow_per_room_profiles` configuration setting in
Synapse's configuration file.
This callback is not called when registering a user, even when setting it through the
[`get_displayname_for_registration`](https://matrix-org.github.io/synapse/latest/modules/password_auth_provider_callbacks.html#get_displayname_for_registration)
module callback.

If multiple modules implement this callback, Synapse runs them all in order.

### `on_deactivation`

_First introduced in Synapse v1.5X.0_

```python
async def on_deactivation(user_id: str, by_admin: bool) -> None:
```

Called after deactivating a local user. The deactivation can be triggered either by the
user themselves or a server admin. The module is passed the Matrix ID of the user that's

> been deactivated, as well as a boolean that is `True` if the deactivation was triggered

Reviewer: ```suggestion
been deactivated, as well as a boolean `by_admin` that is `True` if the deactivation was triggered
```

If multiple modules implement this callback, Synapse runs them all in order.

### `on_profile_update`

> _First introduced in Synapse v1.5X.0_

Author: ```suggestion
_First introduced in Synapse v1.54.0_
```

has been updated, their new profile, as well as a `by_admin` boolean that is `True` if the
update was triggered by a server admin (and `False` otherwise), and a `deactivated`
boolean that is `True` if the update is a result of the user being deactivated.

Note that the `by_admin` boolean is also `True` if the profile change happens as a result
of the user logging in through Single Sign-On, or if a server admin updates their own
profile.

Per-room profile changes do not trigger this callback to be called. Synapse administrators
wishing this callback to be called on every profile change are encouraged to disable
per-room profiles globally using the `allow_per_room_profiles` configuration setting in
Synapse's configuration file.
This callback is not called when registering a user, even when setting it through the
[`get_displayname_for_registration`](https://matrix-org.github.io/synapse/latest/modules/password_auth_provider_callbacks.html#get_displayname_for_registration)
module callback.

If multiple modules implement this callback, Synapse runs them all in order.

### `on_user_deactivation_status_changed`

> _First introduced in Synapse v1.5X.0_

Author: ```suggestion
_First introduced in Synapse v1.54.0_
```
